File: php/LRU/DoubleLinked.php
<?php

require_once 'Node.php';

class DoubleLinked
{

    protected Node $head, $tail;

    protected int $size;

    public function __construct()
    {
        $this->head = new Node(0, 0);
        $this->tail = new Node(0, 0);
        $this->head->next = $this->tail;
        $this->tail->prev = $this->head;
        $this->size = 0;
    }

    public function tailNode()
    {
        return $this->tail;
    }

    public function headNode()
    {
        return $this->head;
    }

    public function size()
    {
        return $this->size;
    }

    public function addLast(Node $node)
    {
        $node->next = $this->tail;
        $node->prev = $this->tail->prev;
        $this->tail->prev->next = $node;
        $this->tail->prev = $node;

        $this->size++;
    }

    public function remove(Node $node)
    {
        $node->prev->next = $node->next;
        $node->next->prev = $node->prev;

        $this->size--;
    }

    public function removeFirst()
    {
        if ($this->head->next === $this->tail) {
            return null;
        }

        $firstNode = $this->head->next;
        $this->remove($firstNode);

        return $firstNode;
    }

}

File: go/lur/double_linked.go
package lur

type Node struct {
	key int
	val string
	next *Node
	prev *Node
}

type DoubleLinked struct {
	head *Node
	tail *Node
	counter int
}

func NewNode(key int, val string, next *Node, prev *Node) *Node {
	return &Node{key, val, next, prev}
}

func NewDoubleLinked () *DoubleLinked {
	d :=  &DoubleLinked{
					NewNode(0, "", nil, nil),
					NewNode(0, "", nil, nil),
						0,
					}

	d.tail.prev = d.head
	d.head.next = d.tail

	return d
}

func (d *DoubleLinked) AddFirst(node *Node)  {
	node.prev = d.head
	node.next = d.head.next
	d.head.next.prev = node
	d.head.next = node

	d.counter++
}

func (d *DoubleLinked) Remove(node *Node) {
	node.next.prev = node.prev
	node.prev.next = node.next
	node = nil

	d.counter--
}

func (d *DoubleLinked) RemoveLast() *Node {
	if d.tail.prev == d.head {
		return nil
	}

	node := d.tail.prev
	d.Remove(node)

	return node
}

